cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/HierarchyTreeNode/index.tsx

```diff
@@ -0,0 +1,338 @@
+      marginLeft: (props) => (props.margin === 0 ? `10px` : `${props.margin}px`),
+    },
+    relative: {
+      position: 'relative',
+    },
+    checkBox: {
+      display: 'flex',
+      width: '100%',
+    },
+    paddingTextField: {
+      paddingLeft: '10px',
+    },
+  };
+});
+
+interface INodeProps {
+  id: number;
+  parentId?: number;
+  children?: INodeProps[];
+  name: string;
+  path?: string[];
+  type: string;
+}
+interface IHierarchyTreeNodeProps extends IHierarchyProps {
+  node: INodeProps;
+  setRecords: (value) => void | React.Dispatch<any>;
+  setTree: (value) => void | React.Dispatch<any>;
+  records: INodeProps[];
+  dropdownOptions: INodeProps[];
+  disabled: boolean;
+}
+
+const HierarchyTreeNode = ({
+  node,
+  setRecords,
+  setTree,
+  onChange,
+  records,
+  dropdownOptions,
+  disabled,
+}: IHierarchyTreeNodeProps) => {
+  const [multiSelectOptions, setMultiSelectOptions] = React.useState([]);
+  const [childVisible, setChildVisiblity] = React.useState(false);
+  const [showMultiSelect, setShowMultiSelect] = React.useState(false);
+  const hasChild = node.children.length !== 0 ? true : false;
+
+  let mrgL = 0;
+  if (hasChild && childVisible) {
+    mrgL += 10;
+  }
+
+  const classes = useStyles({ margin: mrgL });
+
+  const fillMultiSelectOptions = (e) => {
+    setMultiSelectOptions(e);
+  };
+
+  const handleMultiSelectClose = () => {
+    const newRecords = [...records];
+    multiSelectOptions.forEach((item) => {
+      let childrenOfSelected = [];
+
+      const uniqueId = uuidV4();
+
+      if (item.children.length !== 0) {
+        childrenOfSelected = handleChildren(childrenOfSelected, item.children, uniqueId, 0);
+      }
+      const childToBeAdded = {
+        id: uniqueId,
+        path: item.path,
+        parentId: node.id,
+        name: item.name,
+        children: [],
+        type: item.type,
+      };
+      setShowMultiSelect(false);
+      setChildVisiblity(true);
+      newRecords.push(childToBeAdded);
+      if (childrenOfSelected) {
+        childrenOfSelected.forEach((child) => {
+          newRecords.push(child);
+        });
+      }
+    });
+    setTree(flatToTree(newRecords));
+    onChange(JSON.stringify(toJsonHandler(newRecords)));
+    setRecords(newRecords);
+    setMultiSelectOptions([]);
+  };
+
+  const nameChangeHandler = (e) => {
+    const newRecords = [...records];
+    newRecords.forEach((record) => {
+      if (record.id === node.id) {
+        record.name = e.target.value;
+      }
+    });
+    setTree(flatToTree(newRecords));
+    onChange(JSON.stringify(toJsonHandler(newRecords)));
+    setRecords(newRecords);
+  };
+
+  const handleAddNewRecordToParent = () => {
+    const newRecords = [...records];
+    newRecords.push({
+      id: uuidV4(),
+      parentId: node.id,
+      type: 'record',
+      children: [],
+      name: '',
+    });
+    setTree(flatToTree(newRecords));
+    onChange(JSON.stringify(toJsonHandler(newRecords)));
```

Comment:
This can potentially have an empty record if the user closes the plugin without giving a name. Can we call onChange only when the name is added/modified?